build/testing/ui.go

```diff
@@ -2,30 +2,140 @@ package testing
 
 import (
 	"context"
+	"crypto/sha256"
+	"fmt"
 	"os"
+	"path"
 	"time"
 
 	"dagger.io/dagger"
+	"golang.org/x/sync/errgroup"
 )
 
 func UI(ctx context.Context, client *dagger.Client, ui, flipt *dagger.Container) error {
-	test := ui.
-		WithExec([]string{"npx", "playwright", "install", "chromium", "--with-deps"}).
-		WithServiceBinding("flipt", flipt.
-			WithEnvVariable("CI", os.Getenv("CI")).
-			WithEnvVariable("FLIPT_AUTHENTICATION_METHODS_TOKEN_ENABLED", "true").
-			WithEnvVariable("UNIQUE", time.Now().String()).
-			WithExec(nil)).
-		WithEnvVariable("FLIPT_ADDRESS", "http://flipt:8080").
-		WithExec([]string{"npx", "playwright", "test"})
-	_, err := test.ExitCode(ctx)
+	test, err := buildUI(ctx, ui, flipt)
 	if err != nil {
 		return err
 	}
 
-	test.
+	_, err = test.
+		WithExec([]string{"npx", "playwright", "install", "chromium", "--with-deps"}).
+		WithExec([]string{"npx", "playwright", "test"}).
 		Directory("playwright-report").
 		Export(ctx, "playwright-report")
 
-	return nil
+	return err
+}
+
+func Screenshots(ctx context.Context, client *dagger.Client, flipt *dagger.Container) error {
+	src := client.Host().Directory("./ui/", dagger.HostDirectoryOpts{
+		Include: []string{
+			"./package.json",
+			"./package-lock.json",
+			"./playwright.config.ts",
+			"/screenshots/",
+		},
+	})
+
+	contents, err := src.File("package-lock.json").Contents(ctx)
+	if err != nil {
+		return err
+	}
+
+	cache := client.CacheVolume(fmt.Sprintf("node-modules-screenshot-%x", sha256.Sum256([]byte(contents))))
+
+	ui, err := client.Container().From("node:18-bullseye").
+		WithMountedDirectory("/src", src).WithWorkdir("/src").
+		WithMountedCache("/src/node_modules", cache).
+		WithExec([]string{"npm", "install"}).
+		WithExec([]string{"npx", "playwright@1.36.0", "install", "chromium", "--with-deps"}).
```

Comment:
is there a reason to specify the version of playwright instead of using the one used in the UI ITs? (package.json)

feel like this would allow them to diverge

Reply:
I will remove this again. I had a lot of issues with one step installing one version of playwrights chromium and then when I ran node it expected a chromium version at a different path. Chromium was there, just not the version it expected. This was the node modules poisening problem I mentioned yesterday. Building fresh node modules helped, but I was trying to persist them to improve local caching.

Reply:
I removed it with no problems 👍 

I have a feeling at some point this will come back and bite.
But its a local only tool, and nuking the cache works. So, not that worried.